Add explaining-bitcoin-addresses to bitcoin-magazine/bitcoin-explained (#308)
bitcointranscripts/bitcointranscripts: bitcoin-magazine/bitcoin-explained/explaining-bitcoin-addresses.md

> Aaron Van Wirdum: 00:01:45

Reviewer: I think we can start the transcript here

Author: Alright

Sjors Provoost: 00:09:53

No, and so the...

Aaron Van Wirdum: 00:09:54

Okay, so we have 10 symbols, so that means that once you get by the 11th number, at that point you're going to have to reuse symbols you've already used, so you're now going to use combinations. So in our case, that would be, well it's going to get confusing because the first number is a zero, so then the 11th number is the one and the zero.

Sjors Provoost: 00:10:13

Yeah exactly, and there have been different bases in use, right? So base 360 I believe was used by Babylonians, or maybe base 60. And then for computers we tend to use base 2 internally, because chips are either on or off, so it's 0 or 1, so a long series of 0s and 1s. And you can express any number of that now in order to read machine code typically you would use hexadecimal which is base 16 so that is 0 to 9 and then a to F mm-hmm yeah and right and so base 58 is basically this 58 possible characters to express something with.

Aaron Van Wirdum: 00:10:51

Yeah, it's all numbers and there's different ways of expressing a number based on your base. That determines how many symbols you're using.

Sjors Provoost: 00:11:02

> Right, the trade-off here is readability really, because you could represent machine code as normal characters, so the ASCII alphabet, or the ASCII character set is 256 different characters, So that's base 256. But if you've ever done something like print and then the name of a file, your computer will show complete gibberish on the screen and it will start beeping. And the reason it starts beeping is because one of these codes Somewhere in the base 256 is a beep, which actually makes your terminal beep. So it is completely impractical to view a file using base 256, even though there is a character for every of the 256 things there. So that's why you tend to do that in base 16. Hexadecimal is relatively easy to read, but then it's quite long. If you take a public key and you write it as hexadecimal it's a rather long thing to write down but and bass 58 is a little bit shorter so maybe you know it's easier to copy paste perhaps or it's not even easy to read on the phone like bass 58 is pretty terrible because it's uppercase, lowercase, uppercase, lowercase.

Reviewer: ```suggestion
Right, the trade-off here is readability really, because you could represent machine code as normal characters, so the ASCII alphabet, or the ASCII character set is 256 different characters, So that's base 256. But if you've ever done something like print and then the name of a file, your computer will show complete gibberish on the screen and it will start beeping. And the reason it starts beeping is because one of these codes Somewhere in the base 256 is a beep, which actually makes your terminal beep. So it is completely impractical to view a file using base 256, even though there is a character for every of the 256 things there. So that's why you tend to do that in base 16. Hexadecimal is relatively easy to read, but then it's quite long. If you take a public key and you write it as hexadecimal it's a rather long thing to write down but and base 58 is a little bit shorter so maybe you know it's easier to copy paste perhaps or it's not even easy to read on the phone like base 58 is pretty terrible because it's uppercase, lowercase, uppercase, lowercase.
```

Aaron Van Wirdum: 00:17:13

once again all this is is a version number plus this hash represented in base 58. Is that all it is? Yeah. This is all so much simpler than I once thought, Sjors.

Sjors Provoost: 00:17:26

No, it's really simple. And the only mystery that has been solved today, I guess, is, well, what if you only use the public key but that wasn't done using this system so there is no initial letter that would represent trying to do that

Aaron Van Wirdum: 00:17:40

yeah that was never represented in base68

Sjors Provoost: 00:17:42

otherwise probably that would have been version 0 and then all normal addresses might have started with the two who knows okay

Aaron Van Wirdum: 00:17:48

I think for anyone who already knew this which is probably a good chunk of people this is a very boring episode so far but I think it's gonna get better because oh my god because Sjors we now have a new type of address since a year or two, which starts with BC1.

Sjors Provoost: 00:18:09

BC1Q even, usually.

Aaron Van Wirdum: 00:18:11

Yeah, usually, but not always. And we're getting into that, I think. Yep. So what is this all about?

Sjors Provoost: 00:18:16

> Well, that is BEG32 or BECH32 or however you want to pronounce it. And it's been used since SegWit, basically. And again, it is something that doesn't exist on the blockchain, so it's just a convention that wallets can use. This is a, as the name suggests, a base 32 system. Which means you have almost all the letters, and almost all the numbers, minus some ambiguous characters that you don't want to have, because they look too much like numbers or letters.

Reviewer: ```suggestion
Well, that is BECH32 or however you want to pronounce it. And it's been used since SegWit, basically. And again, it is something that doesn't exist on the blockchain, so it's just a convention that wallets can use. This is a, as the name suggests, a base 32 system. Which means you have almost all the letters, and almost all the numbers, minus some ambiguous characters that you don't want to have, because they look too much like numbers or letters.
```

Aaron Van Wirdum: 00:18:46

> Yeah, and I think one of the big differences compared to Base58 is that this time there are no longer uppercase and lowercase letters. There's just any, every letter is only in there once.

Reviewer: I think base58 is preferred but it's base 58 through out up until here.

Author: Interesting, there are 16 places that has base 58, Fixed

Aaron Van Wirdum: 00:20:58

Yeah, so if Litecoin would want to use these kinds of addresses. Maybe they do actually, I don't know.

Sjors Provoost: 00:21:04

Probably, then they might start with LT.

Aaron Van Wirdum: 00:21:06

Exactly. So these first two letters just refer to which currency is this about? What blockchain is this for?

Sjors Provoost: 00:21:12

Yeah, and it can be, I think, a fairly arbitrary number of letters. The idea is that it's separated by a one.

Aaron Van Wirdum: 00:21:17

Oh, it could be more than two letters as well?

Sjors Provoost: 00:21:19

Yeah, I think initially Bitcoin Cash was using a much longer introduction.

Aaron Van Wirdum: 00:21:23

I see, okay.

Sjors Provoost: 00:21:24

So that's pretty arbitrary. Obviously, you want to conserve space, so BC is nice and short and a one, that's a separator, has no value. So if you look at the, what do all the 32 numbers mean, then one is not in it.

Aaron Van Wirdum: 00:21:36

One just means... Skip this. The human readable part is over, now the fun stuff starts. Now the meat and potatoes.

Sjors Provoost: 00:21:43

>  Right. And the fun stuff, it's a little bit easier actually than with Base58, because there's a convention that says if it's... Well, the convention is it starts with the SegWit version, so the first version of SegWit is 0, which in BEC32 is written as Q. And then it's either followed by 20 bytes or 32 bytes. And that is, then it means either it's the public key hash, or it is the script hash. And they're different lengths now because SegWit uses the SHA-256 hash of the script, rather than the RIPMAT-160 hash of the script. So in base58 the script hash is the same length as the public key hash but in segwit they're not the same length. So Simply by looking at how long the address is, you know whether you're paying to a script or you're paying to a public key. Yeah, sure, we don't have to say it.

Reviewer: ```suggestion
 Right. And the fun stuff, it's a little bit easier actually than with Base58, because there's a convention that says if it's... Well, the convention is it starts with the SegWit version, so the first version of SegWit is 0, which in BECH32 is written as Q. And then it's either followed by 20 bytes or 32 bytes. And that is, then it means either it's the public key hash, or it is the script hash. And they're different lengths now because SegWit uses the SHA-256 hash of the script, rather than the RIPMAT-160 hash of the script. So in base58 the script hash is the same length as the public key hash but in segwit they're not the same length. So Simply by looking at how long the address is, you know whether you're paying to a script or you're paying to a public key. Yeah, sure, we don't have to say it.
```

>  Right. And the fun stuff, it's a little bit easier actually than with Base58, because there's a convention that says if it's... Well, the convention is it starts with the SegWit version, so the first version of SegWit is 0, which in BEC32 is written as Q. And then it's either followed by 20 bytes or 32 bytes. And that is, then it means either it's the public key hash, or it is the script hash. And they're different lengths now because SegWit uses the SHA-256 hash of the script, rather than the RIPMAT-160 hash of the script. So in base58 the script hash is the same length as the public key hash but in segwit they're not the same length. So Simply by looking at how long the address is, you know whether you're paying to a script or you're paying to a public key. Yeah, sure, we don't have to say it.

Reviewer: ```suggestion
 Right. And the fun stuff, it's a little bit easier actually than with Base58, because there's a convention that says if it's... Well, the convention is it starts with the SegWit version, so the first version of SegWit is 0, which in BECH32 is written as Q. And then it's either followed by 20 bytes or 32 bytes. And that is, then it means either it's the public key hash, or it is the script hash. And they're different lengths now because SegWit uses the SHA-256 hash of the script, rather than the RIPEMD-160 hash of the script. So in base58 the script hash is the same length as the public key hash but in segwit they're not the same length. So Simply by looking at how long the address is, you know whether you're paying to a script or you're paying to a public key. Yeah, sure, we don't have to say it.
```

Aaron Van Wirdum: 00:28:30

Right.

Sjors Provoost: 00:28:31

So this is why there's a new standard proposed BIP 350 which is called BEC32M and it's actually a very simple change. I think it adds to the all the math it adds one extra number to that math. And then it fixes that particular bug. And everybody's happy.

Aaron Van Wirdum: 00:28:51

So it fixes the bug that the queues don't matter anymore.

Sjors Provoost: 00:28:56

Yeah, you can't just add stuff to it without running into problems.

Aaron Van Wirdum: 00:29:00

But I guess this does mean that wallets that have by now upgraded to support these special SecWit addresses, BEC32 addresses, they now have to upgrade again.

Sjors Provoost: 00:29:11

> That's right. So that's annoying, because it does mean that if your wallet wants to support sending to a taproot address, then it has to make a small change to the BEC32 implementation. And there's some example code on the BIP. It's not a big change because it just adds one number and if you look at the Bitcoin Core implementation it's a fairly simple change that does it. But it does mean that moving forward when you see a back 32 address you have to parse it then see if it's the version 0 or the version 1 and then do things slightly differently but even that is just a very small change right but it is annoying yeah it does mean that especially hardware wallets, you know, with firmware updates could take a while.

Reviewer: ```suggestion
That's right. So that's annoying, because it does mean that if your wallet wants to support sending to a taproot address, then it has to make a small change to the BECH32 implementation. And there's some example code on the BIP. It's not a big change because it just adds one number and if you look at the Bitcoin Core implementation it's a fairly simple change that does it. But it does mean that moving forward when you see a back 32 address you have to parse it then see if it's the version 0 or the version 1 and then do things slightly differently but even that is just a very small change right but it is annoying yeah it does mean that especially hardware wallets, you know, with firmware updates could take a while.
```

Aaron Van Wirdum: 00:29:51

> Right, so we started out with base58 addresses, now we're all starting to use BEC32 addresses. Is this final? Are we going to keep using BEC32 or are you anticipating some other address format somewhere in the future?

Reviewer: ```suggestion
Right, so we started out with base58 addresses, now we're all starting to use BECH32 addresses. Is this final? Are we going to keep using BECH32 or are you anticipating some other address format somewhere in the future?
```
